### PR TITLE
Microsoft Dynamics FetchLeads  - update

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | tba
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

This bugfix/feature has been sponsored by @Webmecanik

[//]: # ( Required: )
#### Description:
Follow up to https://github.com/mautic/mautic/pull/5229 (@npracht  should close old PR)

This PR add the ability for users of Microsoft Dynamics users
- fetch leads from leads Microsoft Dynamics object (contacts and companies are supported so far)
- resolve duplicate contact if contact exists on both sides Mautic and Microsoft Dynamics
  -  we're trying find existed contact and if exists, we just create integration entity and update contact fields based on plugin settings

image

#### Steps to test this PR:

1. Steps to test this PR:
2. Connect your integration (follow documentation)
3. In the feature tab, enable Leads
4. Create a lead in you MSD account
5. Launch command
6. See your lead in mautic contacts
7. Create contacts on both sides Mautic and Microsoft Dynamics
8. Try setup plugin, especially new option Mautic contacts pushed to Dynamics CRM object and Try find exists contact during the push to Dynamics CRM (see image example above)
9. Test if works properly 

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/5674"></a>

### CODE REVIEW

**Comment by kuzmany** [comment]
Label: WIP

**Comment by kuzmany** [comment]
Label: Feature Request

**Comment by kuzmany** [comment]
Label: WIP

**Comment by kuzmany** [comment]
Label: Ready to test

**Comment by npracht** [comment]
I start testing !

**Comment by dbhurley** [comment]
Due to the work being discussed in removing plugins from core and especially extracting the CRM plugins from core I recommend this work be held out of 2.13. This should be part of the discussion when evaluating how CRM plugins are handled, however, additions to Mautic core CRM is being restricted in the immediate as a result. 

_eg. This should live in a `mautic-dynamics` repository instead of being part of the core release.

**Comment by npracht** [comment]
@dbhurley clear. Thanks.

**Comment by npracht** [comment]
@dbhurley i'd suggest that we integrate that to mautic 2.15 as far as it is enhancement of existing. And we'll take out of mautic core when working on Mautic 3 ?

Please confirm.

**Review by npracht — DISMISSED** [review]
All the new features are working. @kuzmany could you please add the documentation ?

**Comment by jmbarneck** [comment]
I just tested this PR and it does not seem to work, here are my results:

[root@cp mauticdev]# php app/console mautic:integration:fetchleads -i Dynamics
Command to fetch contacts from integration
Fetch contacts command is starting

Fetching contacts...
    0 [>---------------------------]
0 contacts were updated and 0 contacts were created

Fetching companies
Fetch companies command is starting
    0 [>---------------------------]
0 companies were updated and 0 companies were created

Updating/creating leads from Mautic to Dynamics
Number of contacts pushed: 0 updated, 0 created, 0 had errors and 0 were ignored (likely duplicates or didn't match field criteria)

[root@cp mauticdev]# php app/console mautic:integration:synccontacts -i Dynamics
Command to fetch contacts from integration
Fetch contacts command is starting

Fetching contacts...
    0 [>---------------------------]
0 contacts were updated and 0 contacts were created

Fetching companies
Fetch companies command is starting
    0 [>---------------------------]
0 companies were updated and 0 companies were created

Updating/creating leads from Mautic to Dynamics
Number of contacts pushed: 0 updated, 0 created, 0 had errors and 0 were ignored (likely duplicates or didn't match field criteria)

It would be great to get this integration working. Let me know if I can provide more information to help.

**Comment by javiergas** [comment]
Hi @kuzmany @npracht 
What do we need / how can I help for this PR to be merged?

**Comment by npracht** [comment]
Hi @javiergas you can test it, tell how is working and not working then Zdeno can finish the work !

File: plugins/MauticCrmBundle/Integration/DynamicsIntegration.php [hunk 1]
```diff
@@ -629,8 +689,8 @@ public function amendLeadDataBeforeMauticPopulate($data, $object = null)
                         $result[] = $entity->getName();
                     }
                     $mauticObjectReference = 'company';
-                } elseif ('contacts' === $object) {
-                    $recordId = $entityData['contactid'];
+                } elseif ('contacts' === $object || 'leads' === $object) {
+                    $recordId = ('contacts' === $object) ? $entityData['contactid'] : $entityData['leadid'];
                     // first try to find integration entity
                     $integrationId = $integrationEntityRepo->getIntegrationsEntityId('Dynamics', $object, 'lead',
```
Comment (javiergas) [inline plugins/MauticCrmBundle/Integration/DynamicsIntegration.php:711]: Hi @npracht, 
I found something very strange that is preventing records by updated from Dynamics:
This line returns nothing but If check what is set to be returned from inside getIntegrationsEntityId() it seems to be OK.

Using PHP 7.2.19-0ubuntu0.18.04.1 (cli) (built: Jun  4 2019 14:48:12) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.19-0ubuntu0.18.04.1, Copyright (c) 1999-2018, by Zend Technologies

**Review by javiergas — CHANGES_REQUESTED** [review]
Records updated in Dynamics are not updated in Mautic

**Comment by florentpetitjean** [comment]
Hello @kuzmany can you give us an update on this ? 
We will need this very soon for a client.

Thanks,

**Comment by kuzmany** [comment]
@javiergas @florentpetitjean  can you test it?

**Review by florentpetitjean — COMMENTED** [review]
I test and it works except for one element : 
When I select lead in specific features  (in the last options, the object to pull data from): 
image
I can't validate the settings as it require a new field in the contact mapping without showing wich one 
image

A part form that  : 

- contacts are synchronised in both way
- The option to push to contact or lead in Dynamics works
- And the contact update if they already exists

**Comment by kuzmany** [comment]
@florentpetitjean  can you re-test with last changes?

**Review by florentpetitjean — DISMISSED** [review]
Tested, all is working fine now, good job !

**Comment by npracht** [comment]
@kuzmany can you add documentation please ?

**Comment by florentpetitjean** [comment]
Hello @kuzmany 
We got this error  after patching this PR, could you look at it ? Thanks !

[2019-11-26 09:40:06] mautic.NOTICE: Mautic\PluginBundle\Exception\ApiErrorException: Dynamics CRM API error: {"code":400,"headers":{"Cache-Control":"no-cache","Allow":"OPTIONS,GET,HEAD,POST","Content-Type":"multipart\/mixed; boundary=batchresponse_018d32cf-ea07-4fc2-a9f6-829726d45db2","Expires":"-1","Server":"Microsoft-IIS\/10.0","x-ms-service-request-id":"ff2b38c8-750d-4164-bbce-5d692fa53682","Strict-Transport-Security":"max-age=31536000; includeSubDomains","REQ_ID":"ff2b38c8-750d-4164-bbce-5d692fa53682","AuthActivityId":"b8ed2998-0152-42b9-a278-8a10e8fa02f5","x-ms-ratelimit-burst-remaining-xrm-requests":"5994","x-ms-ratelimit-time-remaining-xrm-requests":"1,199.87","OData-Version":"4.0","X-Source":"2061881637365871852146011751712502789492542351481987914118417696641642375934148","Public":"OPTIONS,GET,HEAD,POST","Timing-Allow-Origin":"*","X-Powered-By":"ASP.NET","Date":"Tue, 26 Nov 2019 09:40:05 GMT","Content-Length":"1474"},"body":"--batchresponse_018d32cf-ea07-4fc2-a9f6-829726d45db2\r\nContent-Type: multipart\/mixed; boundary=changesetresponse_ec89406b-3523-49f3-89ea-f5154c9b520f\r\n\r\n--changesetresponse_ec89406b-3523-49f3-89ea-f5154c9b520f\r\nContent-Type: application\/http\r\nContent-Transfer-Encoding: binary\r\nContent-ID: 1145\r\n\r\nHTTP\/1.1 400 Bad Request\r\nContent-Type: application\/json; odata.metadata=minimal\r\nOData-Version: 4.0\r\n\r\n{\"error\":{\"code\":\"0x8006088a\",\"message\":\"Requested API Version 'data' was not understood. Please provide in the format \/api\/data\/v.\/ (example: \/api\/data\/v7.2\/)\",\"innererror\":{\"message\":\"Requested API Version 'data' was not understood. Please provide in the format \/api\/data\/v.\/ (example: \/api\/data\/v7.2\/)\",\"type\":\"Microsoft.Crm.CrmHttpException\",\"stacktrace\":\"   at Microsoft.Crm.Extensibility.ODataV4.Routing.CrmODataRoutingConvention.ParseVersion(String versionString, String exceptionmessage)\\r\\n   at Microsoft.Crm.Extensibility.ODataV4.Routing.CrmODataRoutingConvention.GetApiVersionFromRequestUri(Uri requestUri, String& echoVersion, Boolean accessedAsIfd)\\r\\n   at Microsoft.Crm.Extensibility.ODataV4.Routing.CrmODataPathHandler.ValidateApiVersion(Uri serviceRoot)\\r\\n   at Microsoft.Crm.Extensibility.ODataV4.Routing.CrmODataPathHandler.Parse(String serviceRoot, String odataPath, IServiceProvider requestContainer)\"}}}\r\n--changesetresponse_ec89406b-3523-49f3-89ea-f5154c9b520f--\r\n--batchresponse_018d32cf-ea07-4fc2-a9f6-829726d45db2--\r\n"} (uncaught exception) at /home/atmt_arademo/arademo.automation.webmecanik.com/2-15-4-atmt/plugins/MauticCrmBundle/Api/DynamicsApi.php line 58 while running console command `mautic:integration:fetchleads` [] []

**Comment by kuzmany** [comment]
@florentpetitjean last commit should fix push to integration object issue. Please re-test

**Comment by autoize** [comment]
@kuzmany, we notice the same 400 error as @florentpetitjean when testing this pull request on 2.15.3. It occurs after a lead has been updated in Dynamics, and the changes should be synced back to Mautic (in this case Dynamics is set as the system of record).

After applying the change mentioned in this [comment](https://github.com/mautic/mautic/issues/4626#issuecomment-450332944) made by @afonseca08 in #4626, the changes **do** successfully sync over from Integration => Mautic.

However we are still getting the same output in console as mentioned by @florentpetitjean.

```
[example@example go]$ sudo -u apache php app/console -v mautic:integration:fetchleads -i Dynamics
Command to fetch contacts from integration
Fetch contacts command is starting

Fetching contacts...
jackie@example.onmicrosoft.com
standaloneform@example.com
campaignform@example.com
john.doe@example.com
jane.doe@example.com

Number of leads/contacts fetched: 5

Fetching companies
Fetch companies command is starting
example

Number of companies fetched: 1

Updating/creating leads from Mautic to Dynamics
About 1 to update and about 0 to create/update
This could take some time. Please wait until the process is completed
 1/1 [============================] 100%

  [Mautic\PluginBundle\Exception\ApiErrorException]
  Dynamics CRM API error: {"code":400,"headers":{"Cache-Control":"no-cache","Allow":"OPTIONS,GET,HEAD,POST","Content-Type":"multipar
  t\/mixed; boundary=batchresponse_057576c6-1e68-4f46-812c-0863e9315051","Expires":"-1","Server":"Microsoft-IIS\/10.0","x-ms-service
  -request-id":"0013f6c4-7f91-4a54-bc13-ba7295041b9a","Strict-Transport-Security":"max-age=31536000; includeSubDomains","REQ_ID":"00
  13f6c4-7f91-4a54-bc13-ba7295041b9a","AuthActivityId":"640defa0-2778-4dad-aece-6c4a0d5e346a","x-ms-ratelimit-burst-remaining-xrm-re
  quests":"5986","x-ms-ratelimit-time-remaining-xrm-requests":"1,199.56","OData-Version":"4.0","X-Source":"1269921413096212401753427
  1210610494200161661612514419225159121571491722217225213029","Public":"OPTIONS,GET,HEAD,POST","Timing-Allow-Origin":"*","X-Powered-
  By":"ASP.NET","Date":"Thu, 12 Dec 2019 09:33:19 GMT","Content-Length":"5705"},"body":"--batchresponse_057576c6-1e68-4f46-812c-0863
  e9315051\r\nContent-Type: multipart\/mixed; boundary=changesetresponse_5bb4e116-b8c5-4f18-a39b-d10969f0a852\r\n\r\n--changesetresp
  onse_5bb4e116-b8c5-4f18-a39b-d10969f0a852\r\nContent-Type: application\/http\r\nContent-Transfer-Encoding: binary\r\nContent-ID: 7
  7b5bb61-c01c-ea11-a811-000d3a56957f\r\n\r\nHTTP\/1.1 400 Bad Request\r\nREQ_ID: 0013f6c4-7f91-4a54-bc13-ba7295041b9a\r\nContent-Ty
  pe: application\/json; odata.metadata=minimal\r\nOData-Version: 4.0\r\n\r\n{\"error\":{\"code\":\"0x0\",\"message\":\"An error occ
  urred while validating input parameters: Microsoft.OData.ODataException: An unexpected 'StartArray' node was found when reading fr
  om the JSON reader. A 'StartObject' node was expected.\\r\\n   at Microsoft.OData.Json.JsonReaderExtensions.ValidateNodeType(IJson
  Reader jsonReader, JsonNodeType expectedNodeType)\\r\\n   at Microsoft.OData.JsonLight.ODataJsonLightDeserializer.ReadPayloadStart
  Implementation(ODataPayloadKind payloadKind, PropertyAndAnnotationCollector propertyAndAnnotationCollector, Boolean isReadingNeste
  dPayload, Boolean allowEmptyPayload)\\r\\n   at Microsoft.OData.JsonLight.ODataJsonLightDeserializer.ReadPayloadStart(ODataPayload
  Kind payloadKind, PropertyAndAnnotationCollector propertyAndAnnotationCollector, Boolean isReadingNestedPayload, Boolean allowEmpt
  yPayload)\\r\\n   at Microsoft.OData.JsonLight.ODataJsonLightReader.ReadAtStartImplementation()\\r\\n   at Microsoft.OData.ODataRe
  aderCore.ReadImplementation()\\r\\n   at Microsoft.OData.ODataReaderCore.InterceptException[T](Func`1 action)\\r\\n   at System.We
  b.OData.Formatter.Deserialization.ODataReaderExtensions.ReadResourceOrResourceSet(ODataReader reader)\\r\\n   at System.Web.OData.
  Formatter.Deserialization.ODataResourceDeserializer.Read(ODataMessageReader messageReader, Type type, ODataDeserializerContext rea
  dContext)\\r\\n   at System.Web.OData.Formatter.ODataMediaTypeFormatter.ReadFromStream(Type type, Stream readStream, HttpContent c
  ontent, IFormatterLogger formatterLogger)\",\"innererror\":{\"message\":\"An error occurred while validating input parameters: Mic
  rosoft.OData.ODataException: An unexpected 'StartArray' node was found when reading from the JSON reader. A 'StartObject' node was
   expected.\\r\\n   at Microsoft.OData.Json.JsonReaderExtensions.ValidateNodeType(IJsonReader jsonReader, JsonNodeType expectedNode
  Type)\\r\\n   at Microsoft.OData.JsonLight.ODataJsonLightDeserializer.ReadPayloadStartImplementation(ODataPayloadKind payloadKind,
   PropertyAndAnnotationCollector propertyAndAnnotationCollector, Boolean isReadingNestedPayload, Boolean allowEmptyPayload)\\r\\n
   at Microsoft.OData.JsonLight.ODataJsonLightDeserializer.ReadPayloadStart(ODataPayloadKind payloadKind, PropertyAndAnnotationColle
  ctor propertyAndAnnotationCollector, Boolean isReadingNestedPayload, Boolean allowEmptyPayload)\\r\\n   at Microsoft.OData.JsonLig
  ht.ODataJsonLightReader.ReadAtStartImplementation()\\r\\n   at Microsoft.OData.ODataReaderCore.ReadImplementation()\\r\\n   at Mic
  rosoft.OData.ODataReaderCore.InterceptException[T](Func`1 action)\\r\\n   at System.Web.OData.Formatter.Deserialization.ODataReade
  rExtensions.ReadResourceOrResourceSet(ODataReader reader)\\r\\n   at System.Web.OData.Formatter.Deserialization.ODataResourceDeser
  ializer.Read(ODataMessageReader messageReader, Type type, ODataDeserializerContext readContext)\\r\\n   at System.Web.OData.Format
  ter.ODataMediaTypeFormatter.ReadFromStream(Type type, Stream readStream, HttpContent content, IFormatterLogger formatterLogger)\",
  \"type\":\"Microsoft.Crm.CrmHttpException\",\"stacktrace\":\"   at Microsoft.Crm.Extensibility.OData.CrmODataUtilities.ValidateInp
  utParameters(ModelStateDictionary controllerModelState)\\r\\n   at Microsoft.Crm.Extensibility.OData.EntityController.PatchEntityI
  mplementation(String& entityName, String key, EdmEntityObject entityDelta)\\r\\n   at Microsoft.PowerApps.CoreFramework.ActivityLo
  ggerExtensions.Execute[TResult](ILogger logger, EventId eventId, ActivityType activityType, Func`1 func, IEnumerable`1 additionalC
  ustomProperties)\\r\\n   at Microsoft.Xrm.Telemetry.XrmTelemetryExtensions.Execute[TResult](ILogger logger, XrmTelemetryActivityTy
  pe activityType, Func`1 func)\\r\\n   at lambda_method(Closure , Object , Object[] )\\r\\n   at System.Web.Http.Controllers.Reflec
  tedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)\\r\\n   a
  t System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 ar
  guments, CancellationToken cancellationToken)\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r
  \\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.Han
  dleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsy
  ncCore>d__0.MoveNext()\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime
  .ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebugge
  rNotification(Task task)\\r\\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()\\r\\n--- End of s
  tack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo
  .Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at Syst
  em.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()\"}}}\r\n--changesetresponse_5bb4e116-b8c5-4f18-a39b-d10
  969f0a852--\r\n--batchresponse_057576c6-1e68-4f46-812c-0863e9315051--\r\n"}


Exception trace:
 () at /var/www/html/example.com/public_html/go/plugins/MauticCrmBundle/Api/DynamicsApi.php:58
 MauticPlugin\MauticCrmBundle\Api\DynamicsApi->request() at /var/www/html/example.com/public_html/go/plugins/MauticCrmBundle/Api/DynamicsApi.php:207
 MauticPlugin\MauticCrmBundle\Api\DynamicsApi->createLeads() at /var/www/html/example.com/public_html/go/plugins/MauticCrmBundle/Api/DynamicsApi.php:223
 MauticPlugin\MauticCrmBundle\Api\DynamicsApi->updateLeads() at /var/www/html/example.com/public_html/go/plugins/MauticCrmBundle/Integration/DynamicsIntegration.php:1062
 MauticPlugin\MauticCrmBundle\Integration\DynamicsIntegration->pushLeads() at /var/www/html/example.com/public_html/go/app/bundles/PluginBundle/Command/FetchLeadsCommand.php:258
 Mautic\PluginBundle\Command\FetchLeadsCommand->execute() at /var/www/html/example.com/public_html/go/vendor/symfony/console/Command/Command.php:242
 Symfony\Component\Console\Command\Command->run() at /var/www/html/example.com/public_html/go/vendor/symfony/console/Application.php:861
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/example.com/public_html/go/vendor/symfony/console/Application.php:193
 Symfony\Component\Console\Application->doRun() at /var/www/html/example.com/public_html/go/vendor/symfony/framework-bundle/Console/Application.php:84
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/html/example.com/public_html/go/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at /var/www/html/example.com/public_html/go/app/console:41
```

**Comment by RCheesley** [comment]
Hi @kuzmany is there any update on this at all? Are we just pending tests and documentation or do we have other work to be done before this can be merged?

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@kuzmany a Travis fix here

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
Still has outstanding test coverage issues, I am willing to bump this one more feature release to Mautic 4 and if we have not merged it by then we will close the PR.

**Comment by RCheesley** [comment]
Conflicts & rebase needed. Pinging @mautic/dynamics-tiger-team - it's an old PR so would be good to check if it's still required!

**Comment by escopecz** [comment]
@kuzmany could you please check if this issue was fixed with https://github.com/mautic/mautic/pull/13208?

**Comment by RCheesley** [comment]
I'm going to close this PR given the lack of response over the last years. If someone wishes to pick it up in the future please ping us to re-open the PR. It will need to be rebased to the latest branch.

**Pull request closed without merging**